rust/uefi/linux-bootloader/src/companions.rs

```diff
@@ -0,0 +1,158 @@
+use crate::cpio::{pack_cpio, Cpio};
+use alloc::{string::ToString, vec::Vec};
+use uefi::{
+    cstr16,
+    fs::{Path, PathBuf},
+    prelude::BootServices,
+    proto::device_path::{
+        text::{AllowShortcuts, DisplayOnly},
+        DevicePath,
+    },
+    CString16,
+};
+
+/// Locate files with ASCII filenames and matching the suffix passed as a parameter.
+/// Returns a list of their paths.
+pub fn find_files(
+    fs: &mut uefi::fs::FileSystem,
+    search_path: &Path,
+    suffix: &str,
+) -> uefi::Result<Vec<PathBuf>> {
+    let mut results = Vec::new();
+
+    for maybe_entry in fs.read_dir(search_path).unwrap() {
+        let entry = maybe_entry?;
+        if entry.is_regular_file() {
+            let fname = entry.file_name();
+            if fname.is_ascii() && fname.to_string().ends_with(suffix) {
+                let mut full_path = CString16::from(search_path.to_cstr16());
+                full_path.push_str(cstr16!("\\"));
+                full_path.push_str(fname);
+                results.push(full_path.into());
+            }
+        }
+    }
+
+    Ok(results)
+}
+
+/// Returns the "default" drop-in directory if it exists.
+/// This will be in general $loaded_image_path.extra/
+pub fn get_default_dropin_directory(
+    boot_services: &BootServices,
+    loaded_image_file_path: &DevicePath,
+    fs: &mut uefi::fs::FileSystem,
+) -> uefi::Result<Option<PathBuf>> {
+    // We could use LoadedImageDevicePath to get the full device path
+    // and perform replacement of the last node before END_ENTIRE
+    // by another node containing the filename + .extra
+    // But this is as much tedious as performing a conversion to string
+    // then opening the root directory and finding the new directory.
+    let mut target_directory = loaded_image_file_path
+        .to_string(boot_services, DisplayOnly(false), AllowShortcuts(false))
+        // This is the Result-level error
+        .expect("Failed to obtain the string representation of the loaded image file path");
+    target_directory.push_str(cstr16!(".extra"));
+
+    return Ok(fs
+        .metadata(target_directory.as_ref())
+        .ok()
+        .and_then(|metadata| {
+            if metadata.is_directory() {
+                Some(PathBuf::from(target_directory))
+            } else {
+                None
+            }
```

Comment:
So interestingly, this doesn't work because:
```
56  |       Ok(fs
    |  _____--_^
    | |     |
    | |     arguments to this enum variant are incorrect
57  | |         .metadata(target_directory.as_ref())
58  | |         .ok()
59  | |         .and_then(|metadata| {
...   |
62  | |                 .then_some(|| PathBuf::from(target_directory))
63  | |         }))
    | |__________^ expected `Option<PathBuf>`, found `Option<{closure@companions.rs:62:28}>`
    |
```

So I don't know if I'm fucking up somewhere but I'm not so sure if there is not a closure capturing problem.

Reply:
Fixed, it was `then`, not `then_some`.